Use the GRADLE_PLUGIN_API_VERSION_ATTRIBUTE to set the required minim…

hfhbd opened the pull request:
…um version of Gradle.

It available since 7.0, but Gradle 8.7 improved the error message: https://github.com/gradle/gradle/issues/24609

---

- [ ] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

hfhbd committed:
Use the GRADLE_PLUGIN_API_VERSION_ATTRIBUTE to set the required minimum version of Gradle.

It available since 7.0, but Gradle 8.7 improved the error message: https://github.com/gradle/gradle/issues/24609

DennisTsar commented:
Just FYI this doesn't actually validate/show the message when using an unsupported Gradle version as Gradle selects the `runtimeElements` variant, not `apiElements` where the attribute is set .

hfhbd commented:
Gradle consumes both resolvable configurations and validates both of them. Let me add a test.

We also use api to fail compilation of consumers of the licensee plugin as a regular library (when creating a new plugin on top of it).

DennisTsar commented:
I can use Licensee 1.14.1 with Gradle 8.9 and get no errors, even though it declares a minimum version of 8.12.

hfhbd commented:
@DennisTsar Sorry, you are right: https://docs.gradle.org/current/userguide/java_library_plugin.html#consumable_configurations The apiElements only contains the `api` dependencies. I noticed the bug/the behavior in a more complex Gradle plugin that also uses `api` elements, so it did work with `apiElements` but the right way is to use `runtimeElements` because it contains `api` AND `implementation` dependencies, so Gradle will always consume the `runtimeElements` when running a Gradle plugin.

I added a test and also fixed it: #542

DennisTsar commented:
I'm not sure if it's a Gradle bug, but even setting the attribute on both `apiElements` and  `runtimeElements` isn't enough to get the nicer error message from https://github.com/gradle/gradle/issues/24609. In Gradle's tests, they set it on [all consumable configurations](https://github.com/gradle/gradle/blob/e6c411193dd81b19a06e469d32529c75b9fda983/platforms/extensibility/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/variants/GradlePluginWithVariantsPublicationIntegrationTest.groovy#L162), which does show the improved error.

hfhbd commented:
I do get the correct (and minified) error message using `runtimeElements` only: https://github.com/hfhbd/licensee/blob/1497de3b25d739fc4868d9b8be4efae8a572f15d/src/test/kotlin/app/cash/licensee/LicenseePluginFixtureTest.kt#L272

I only added `apiElements` (back) to also add plugin authors support.

A java library only has 4 variants by default: apiElements, runtimeElements, sources and javadoc. The latter two do not require the gradle version attributes.

DennisTsar commented:
I believe that's the old error message? The new one is
> Plugin ... requires at least Gradle 9.0. This build uses Gradle 8.9.

hfhbd commented:
Yeah, true... IMHO this is a bug at Gradle, there is no need to validate the sources or javadoc variants.
